bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java

```diff
@@ -0,0 +1,463 @@
+import org.openhab.binding.ecovacs.internal.api.model.CleanLogRecord;
+import org.openhab.binding.ecovacs.internal.api.model.DeviceCapability;
+import org.openhab.binding.ecovacs.internal.api.util.XPathUtils;
+import org.openhab.core.io.net.http.TrustAllTrustManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xmlpull.v1.XmlPullParser;
+
+import com.google.gson.Gson;
+
+/**
+ * @author Danny Baumann - Initial contribution
+ */
+@NonNullByDefault
+public class EcovacsXmppDevice implements EcovacsDevice {
+    private final Logger logger = LoggerFactory.getLogger(EcovacsXmppDevice.class);
+
+    private final Device device;
+    private final DeviceDescription desc;
+    private final EcovacsApiImpl api;
+    private final Gson gson;
+    private @Nullable IncomingMessageHandler messageHandler;
+    private @Nullable PingHandler pingHandler;
+    private @Nullable XMPPTCPConnection connection;
+    private @Nullable Jid ownAddress;
+    private @Nullable Jid targetAddress;
+
+    EcovacsXmppDevice(Device device, DeviceDescription desc, EcovacsApiImpl api, Gson gson) {
+        this.device = device;
+        this.desc = desc;
+        this.api = api;
+        this.gson = gson;
+    }
+
+    @Override
+    public String getId() {
+        return device.getDid();
+    }
+
+    @Override
+    public String getSerialNumber() {
+        return device.getName();
+    }
+
+    @Override
+    public String getModelName() {
+        return desc.modelName;
+    }
+
+    @Override
+    public boolean hasCapability(DeviceCapability cap) {
+        return desc.capabilities.contains(cap);
+    }
+
+    @Override
+    public <T> T sendCommand(IotDeviceCommand<T> command) throws EcovacsApiException {
+        IncomingMessageHandler handler = this.messageHandler;
+        XMPPConnection conn = this.connection;
+        Jid from = this.ownAddress;
+        Jid to = this.targetAddress;
+        if (handler == null || conn == null || from == null || to == null) {
+            throw new EcovacsApiException("Not connected to device");
+        }
+
+        try {
+            // Devices sometimes send no answer to commands for unknown reasons. Ecovacs'
+            // app employs a similar retry mechanism, so this seems to be 'normal'.
+            for (int retry = 0; retry < 3; retry++) {
+                DeviceCommandIQ request = new DeviceCommandIQ(command, from, to);
+                CommandResponseHolder responseHolder = new CommandResponseHolder();
+                handler.registerPendingCommand(request.id, responseHolder);
+
+                logger.trace("{}: sending command {}, retry {}", getSerialNumber(),
+                        command.getName(ProtocolVersion.XML), retry);
+                synchronized (responseHolder) {
+                    conn.sendIqRequestAsync(request);
+                    responseHolder.wait(1500);
+                }
+                handler.unregisterPendingCommand(request.id);
```

Comment:
That smells like something you would want to do in a `finally` block to prevent leaks.

Reply:
Good point, will fix this.